File: 146.cpp
// https://leetcode.com/problems/lru-cache/

class LRUCache{
public:
    using KeyType = int;
    using ValueType = int;
    using NodeType = std::pair<KeyType, ValueType>;
    using CacheType = std::list<NodeType>;
    
    LRUCache(int capacity):size_(0), capacity_(capacity) {
    }
    
    ValueType get(KeyType key) {
        auto itr = index_.find(key);
        if (itr == index_.cend()) {
            return -1;
        } else {
            setMostRecent(itr->second);
            return itr->second->second;
        }
    }
    
    void set(KeyType key, ValueType value) {
        auto itr = index_.find(key);
        // not existed
        if (itr == index_.end()) {
            // capacity full
            if (size_ == capacity_) {
                int oldestKey = cache_.front().first;
                index_.erase(oldestKey);
                cache_.pop_front();
            } else {
                ++size_;
            }
            cache_.emplace_back(key, value);
            index_.insert(std::make_pair(key, prev(cache_.end())));
        } else {
            // update old value
            itr->second->second = value;
            setMostRecent(itr->second);
        }
    }
private:
    void setMostRecent(CacheType::iterator itr) {
        cache_.splice(std::end(cache_), cache_, itr);
    }
    int size_;
    int capacity_;

    
    CacheType cache_;
    std::unordered_map<KeyType, CacheType::iterator> index_;
};

File: golang/146.go
package main

type LRUCache struct {
	Data     *list.List
	Index    map[int]*list.Element
	Num      int
	Capacity int
}
type Pair struct {
	Key   int
	Value int
}

func Constructor(capacity int) LRUCache {
	cache := LRUCache{
		list.New(),
		make(map[int]*list.Element),
		0,
		capacity,
	}
	return cache
}
func (this *LRUCache) Get(key int) int {
	elem, ok := this.Index[key]
	if ok == false {
		return -1
	}
	this.SetMostRecent(elem)
	return elem.Value.(Pair).Value
}
func (this *LRUCache) Put(key int, value int) {
	elem, ok := this.Index[key]
	if ok == false {
		// full
		if this.Num == this.Capacity {
			delete(this.Index, this.Data.Front().Value.(Pair).Key)
			this.Data.Remove(this.Data.Front())
		} else {
			this.Num++
		}
		this.Data.PushBack(Pair{key, value})
		this.Index[key] = this.Data.Back()
	} else {
		elem.Value = Pair{key, value}
		this.SetMostRecent(elem)
	}
}
func (this *LRUCache) SetMostRecent(elem *list.Element) {
	d := this.Data
	d.MoveToBack(elem)
}
